src/tqec/computation/block_graph.py

```diff
@@ -414,6 +414,58 @@ def from_dae_file(filename: str | pathlib.Path, graph_name: str = "") -> BlockGr
 
         return read_block_graph_from_dae_file(filename, graph_name)
 
+    def relabel_cubes(self, label_mapping: Mapping[Position3D | str, str]) -> None:
+        """Relabel cubes in the block graph.
+
+        This method updates the labels of cubes in the graph, based on a mapping
+        from either a cube position or its existing label to a new label.
+
+        Args:
+            label_mapping: A mapping from either Position3D or current cube label (str)
+                to the new label to assign.
+
+        Raises:
+            TQECException: If a cube is not found for the given key, if a port label
+                is reused, or if the new label conflicts with existing port labels.
+        """
+        port_labels = {cube.label for cube in self.cubes if cube.is_port}
+        assigned_new_labels: set[str] = set()
+
+        for key, new_label in label_mapping.items():
+            if not new_label:
+                raise TQECException("New label must be non-empty.")
+
+            if new_label in port_labels:
+                raise TQECException(
+                    f"The label '{new_label}' is already used by a port and cannot be reassigned."
+                )
+            if new_label in assigned_new_labels:
+                raise TQECException(
+                    f"The label '{new_label}' is reused multiple times in this relabeling."
+                )
```

Comment:
These might be all checked before the loop. It is up to preference, here is an example of code that might be used instead.

```py
from collections import Counter
from itertools import chain

class BlockGraph:
    # ....
    def relabel_cubes(self, label_mapping: Mapping[Position3D | str, str]) -> None:
        # ....
        labels = Counter(chain.from_iterables((port_labels, label_mapping.values())))
        duplicated_labels = {label for label, count in labels.items() if count > 1}
        if duplicated_labels:
            raise TQECException(f"Found duplicated label(s): {duplicated_labels}.")
```

It is up to you to decide what you prefer, I am OK with both approaches.